### PR TITLE
Improve error message for BranchPythonOperator when no task_id to follow

### LINKED ISSUE #18454: BranchPythonOperator intermittently fails to find its downstream task ids
### Apache Airflow version

2.2.0b2 (beta snapshot)

### Operating System

debian (docker)

### Versions of Apache Airflow Providers

n/a

### Deployment

Astronomer

### Deployment details

`astro dev start` with dockerfile:

```
FROM quay.io/astronomer/ap-airflow-dev:2.2.0-buster-44011
```

And dags:
- [simple_branch](https://gist.github.com/MatrixManAtYrService/bb7eac11d2a60c840b02f5fc7c4e864f#file-simple_branch-py)
- [branch](https://gist.github.com/MatrixManAtYrService/bb7eac11d2a60c840b02f5fc7c4e864f#file-branch-py)

### What happened

Spamming the "manual trigger" button on `simple_branch` caused no trouble:

good

But doing the same on `branch` caused intermittent failures:

bad

task logs:
```
*** Reading local file: /usr/local/airflow/logs/branch/decide_if_you_like_it/2021-09-23T01:31:50.261287+00:00/2.log
[2021-09-23, 01:41:27 UTC] {taskinstance.py:1034} INFO - Dependencies all met for <TaskInstance: branch.decide_if_you_like_it manual__2021-09-23T01:31:50.261287+00:00 [queued]>
[2021-09-23, 01:41:27 UTC] {taskinstance.py:1034} INFO - Dependencies all met for <TaskInstance: branch.decide_if_you_like_it manual__2021-09-23T01:31:50.261287+00:00 [queued]>
[2021-09-23, 01:41:27 UTC] {taskinstance.py:1232} INFO - 
--------------------------------------------------------------------------------
[2021-09-23, 01:41:27 UTC] {taskinstance.py:1233} INFO - Starting attempt 2 of 2
[2021-09-23, 01:41:27 UTC] {taskinstance.py:1234} INFO - 
--------------------------------------------------------------------------------
[2021-09-23, 01:41:27 UTC] {taskinstance.py:1253} INFO - Executing <Task(BranchPythonOperator): decide_if_you_like_it> on 2021-09-23 01:31:50.261287+00:00
[2021-09-23, 01:41:27 UTC] {standard_task_runner.py:52} INFO - Started process 8584 to run task
[2021-09-23, 01:41:27 UTC] {standard_task_runner.py:76} INFO - Running: ['airflow', 'tasks', 'run', 'branch', 'decide_if_you_like_it', 'manual__2021-09-23T01:31:50.261287+00:00', '--job-id', '216', '--raw', '--subdir', 'DAGS_FOLDER/branch.py', '--cfg-path', '/tmp/tmp82f81d8w', '--error-file', '/tmp/tmpco74n1v0']
[2021-09-23, 01:41:27 UTC] {standard_task_runner.py:77} INFO - Job 216: Subtask decide_if_you_like_it
[2021-09-23, 01:41:27 UTC] {logging_mixin.py:109} INFO - Running <TaskInstance: branch.decide_if_you_like_it manual__2021-09-23T01:31:50.261287+00:00 [running]> on host 992912b473f9
[2021-09-23, 01:41:27 UTC] {taskinstance.py:1406} INFO - Exporting the following env vars:
AIRFLOW_CTX_DAG_OWNER=airflow
AIRFLOW_CTX_DAG_ID=branch
AIRFLOW_CTX_TASK_ID=decide_if_you_like_it
AIRFLOW_CTX_EXECUTION_DATE=2021-09-23T01:31:50.261287+00:00
AIRFLOW_CTX_DAG_RUN_ID=manual__2021-09-23T01:31:50.261287+00:00
[2021-09-23, 01:41:27 UTC] {python.py:152} INFO - Done. Returned value was: None
[2021-09-23, 01:41:27 UTC] {skipmixin.py:143} INFO - Following branch None
[2021-09-23, 01:41:27 UTC] {taskinstance.py:1680} ERROR - Task failed with exception
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/airflow/models/taskinstance.py", line 1315, in _run_raw_task
    self._prepare_and_execute_task_with_callbacks(context, task)
  File "/usr/local/lib/python3.9/site-packages/airflow/models/taskinstance.py", line 1437, in _prepare_and_execute_task_with_callbacks
    result = self._execute_task(context, task_copy)
  File "/usr/local/lib/python3.9/site-packages/airflow/models/taskinstance.py", line 1493, in _execute_task
    result = execute_callable(context=context)
  File "/usr/local/lib/python3.9/site-packages/airflow/operators/python.py", line 181, in execute
    self.skip_all_except(context['ti'], branch)
  File "/usr/local/lib/python3.9/site-packages/airflow/models/skipmixin.py", line 147, in skip_all_except
    branch_task_ids = set(branch_task_ids)
TypeError: 'NoneType' object is not iterable
[2021-09-23, 01:41:27 UTC] {taskinstance.py:1261} INFO - Marking task as FAILED. dag_id=branch, task_id=decide_if_you_like_it, execution_date=20210923T013150, start_date=20210923T014127, end_date=20210923T014127
[2021-09-23, 01:41:27 UTC] {standard_task_runner.py:88} ERROR - Failed to execute job 216 for task decide_if_you_like_it
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/airflow/task/task_runner/standard_task_runner.py", line 85, in _start_by_fork
    args.func(args, dag=self.dag)
  File "/usr/local/lib/python3.9/site-packages/airflow/cli/cli_parser.py", line 48, in command
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/airflow/utils/cli.py", line 92, in wrapper
    return f(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/airflow/cli/commands/task_command.py", line 292, in task_run
    _run_task_by_selected_method(args, dag, ti)
  File "/usr/local/lib/python3.9/site-packages/airflow/cli/commands/task_command.py", line 107, in _run_task_by_selected_method
    _run_raw_task(args, ti)
  File "/usr/local/lib/python3.9/site-packages/airflow/cli/commands/task_command.py", line 180, in _run_raw_task
    ti._run_raw_task(
  File "/usr/local/lib/python3.9/site-packages/airflow/utils/session.py", line 70, in wrapper
    return func(*args, session=session, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/airflow/models/taskinstance.py", line 1315, in _run_raw_task
    self._prepare_and_execute_task_with_callbacks(context, task)
  File "/usr/local/lib/python3.9/site-packages/airflow/models/taskinstance.py", line 1437, in _prepare_and_execute_task_with_callbacks
    result = self._execute_task(context, task_copy)
  File "/usr/local/lib/python3.9/site-packages/airflow/models/taskinstance.py", line 1493, in _execute_task
    result = execute_callable(context=context)
  File "/usr/local/lib/python3.9/site-packages/airflow/operators/python.py", line 181, in execute
    self.skip_all_except(context['ti'], branch)
  File "/usr/local/lib/python3.9/site-packages/airflow/models/skipmixin.py", line 147, in skip_all_except
    branch_task_ids = set(branch_task_ids)
TypeError: 'NoneType' object is not iterable
[2021-09-23, 01:41:27 UTC] {local_task_job.py:154} INFO - Task exited with return code 1
[2021-09-23, 01:41:27 UTC] {local_task_job.py:264} INFO - 0 downstream tasks scheduled from follow-on schedule check
```

### What you expected to happen

`BranchPythonOperator` should create tasks that always succeed

### How to reproduce

Keep clicking manual executions of the dag called `branch` until you've triggered ten or so.  At least one of them will fail with the error:

```
TypeError: 'NoneType' object is not iterable
```

### Anything else

I've also seen this behavior in image: `quay.io/astronomer/ap-airflow-dev:2.1.4-buster-44000`

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by eladkal**
closes: https://github.com/apache/airflow/issues/18454

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

**Comment by uranusjr** [comment]
Instead of explicitly forbidding this, would it be reasonable to instead handle the None case and skip all the downstreams instead?

**Comment by eladkal** [comment]
> Instead of explicitly forbidding this, would it be reasonable to handle the None case and skip all the downstreams instead?

When would None be desired?

**Comment by josh-fell** [comment]
> > Instead of explicitly forbidding this, would it be reasonable to handle the None case and skip all the downstreams instead?
> 
> When would None be desired?

I could understand users may not want the DAG to fail outright if the branching section is not critical path like with some feature engineering use cases I've seen.  It may be more intuitive to not have to create an empty branch to simply continue with the pipeline if an output doesn't need to be tracked.

Could this be a configurable mode: fail or skip if no branches found?

**Comment by eladkal** [comment]
> I could understand users may not want the DAG to fail outright if the branching section is not critical path like with some feature engineering use cases I've seen. It may be more intuitive to not have to create an empty branch to simply continue with the pipeline if an output doesn't need to be tracked. An edge case for sure with an obvious workaround.
> 
> Could this be a configurable mode: fail or skip if no branches found?

This sounds more like a wrong usage of branching. For conditional junctions we have ShortCircuitOperator
but i'm OK with @uranusjr suggestion if this is prefered.

**Comment by josh-fell** [comment]
Perhaps but the ShortCircuitOperator can be redundant here if we're skipping downstream tasks with this operator now. Seconded on @uranusjr's suggestion as well FWIW.

**Comment by MatrixManAtYrService** [comment]
I had the same thought as @uranusjr, skipping all downstream tasks seems like an obvious way to interpret `None` to me, like an `if` statement with no `else`.

**Comment by eladkal** [comment]
@uranusjr correct me if I'm wrong but this requires rewriting [skip_all_except](https://github.com/apache/airflow/blob/958b679dae6bbd9def7c60191c3a7722ce81382a/airflow/models/skipmixin.py#L134)
The way I see it I can't just blindly skip all downstream tasks as unlike `ShortCircuitOperator` branching does respect trigger rules so I need to skip all downstream tasks until a join/junction task is found.
Since my input is None it's up to the code to search if such junction exists.
So I need to set difference between two cases:

 If I have a case of:

            # branch  ----->  task2
            #   \           
            #    \        
            #       task1
            
Then I need to execute [skip](https://github.com/apache/airflow/blob/958b679dae6bbd9def7c60191c3a7722ce81382a/airflow/models/skipmixin.py#L66)
but If I have a case like:

            # branch  ----->  task2   <----- task3
            #   \           
            #    \        
            #       task1
           
 Then I need to run `skip_all_except(context['ti'],['task2])` because task2 may still run depends on its `trigger_rule`.

**Comment by ashb** [comment]
If you want to skip everything you can `raise AirflowSkipException` I think. (that would have a slightly different behaviour of also skipping the branch op itself, unless we caught that in this op and turned it in to skip all downstream)

That's all a separate pr/feature though

**Comment by eladkal** [comment]
> If you want to skip everything you can `raise AirflowSkipException` I think. (that would have a slightly different behaviour of also skipping the branch op itself, unless we caught that in this op and turned it in to skip all downstream)

So this is a possible route. We will need to keep the change in `SkipMixin` itself to avoid changing all the branch operators.

> That's all a separate pr/feature though

I agree. This PR is just replacing the error message to a more meaningful one. There is no change in functionality here.
I'll work on trying to solve the issue from the root for a future PR.

File: airflow/operators/python.py [hunk 1]
```diff
@@ -178,6 +178,8 @@ class BranchPythonOperator(PythonOperator, SkipMixin):
 
     def execute(self, context: Dict):
         branch = super().execute(context)
+        if branch is None:
+            raise AirflowException("Branch callable must return at least one task_id to follow")
```
Comment (uranusjr) [inline airflow/operators/python.py]: Kind of wondering whether we should check for all non-str values. Or even check the value is a valid task ID? Something like…

```python
if isinstance(branch, str):
    branches = [branch]
try:
    branches = list(branch)  # Check iterability.
except TypeError:
    raise AirflowException("Branch callable must return either a task ID or a list of IDs")
valid_task_ids = set(context["dag"].task_ids)
if not any(b in valid_task_ids for b in branches):
    raise AirflowException("Branch callable must return at least one task ID to follow")
```

**Review by uranusjr — APPROVED** [review]

File: airflow/operators/python.py (hunk 1 above)
Comment (eladkal) [reply airflow/operators/python.py]: yeah this is something I had in mind - added it in last commit
In general this is something we should have in skipmixin as it's logic shared by all branch operators.

File: airflow/operators/python.py [hunk 2]
```diff
@@ -178,6 +178,16 @@ class BranchPythonOperator(PythonOperator, SkipMixin):
 
     def execute(self, context: Dict):
         branch = super().execute(context)
+        # TODO: The logic should be moved to SkipMixin to be available to all branch operators.
+        if not isinstance(branch, (str, List)):
+            raise AirflowException("Branch callable must return either a task ID or a list of IDs")
+        branches = {branch}
```
Comment (uranusjr) [inline airflow/operators/python.py]:
```suggestion
        if isinstance(branch, str):
            branches = {branch}
        elif isinstance(branch, list):
            branches = set(branch)
        else:
            raise AirflowException("Branch callable must return either a task ID or a list of IDs")
```

Otherwise if `branch` is a list (e.g. `["task_a", "task_b"]`), `branches` would become a set containing one list of strings (`{["task_a", "task_b"]}`). We need a set of strings (`{"task_a", "task_b"}`).

File: airflow/operators/python.py [hunk 3]
```diff
@@ -178,6 +178,19 @@ class BranchPythonOperator(PythonOperator, SkipMixin):
 
     def execute(self, context: Dict):
         branch = super().execute(context)
+        # TODO: The logic should be moved to SkipMixin to be available to all branch operators.
+        if isinstance(branch, str):
```
Comment (potiuk) [inline airflow/operators/python.py:182]: Should we also change SkipMixin while we are changing it (or will it be a separate PR) ? (per the comment above)

Comment (eladkal) [reply airflow/operators/python.py:182]: Seperate PR.
I will work on it. We want also to change the functionality so it will accept None and skip all downstream tasks (untill a junction/join task).

**Review by potiuk — APPROVED** [review]

File: airflow/operators/python.py [hunk 4]
```diff
@@ -178,6 +178,19 @@ class BranchPythonOperator(PythonOperator, SkipMixin):
...
+            branches = {branch}
+        elif isinstance(branch, list):
+            branches = set(branch)
+        else:
+            raise AirflowException("Branch callable must return either a task ID or a list of IDs")
+        valid_task_ids = set(context["dag"].task_ids)
+        invalid_task_ids = branches - valid_task_ids
+        if invalid_task_ids:
+            raise AirflowException(
+                f"Branch callable must valid task_ids. Invalid tasks found: {invalid_task_ids}"
```
Comment (edwardwang888) [inline airflow/operators/python.py:192]: @eladkal Did you mean something like
`Branch callable must *return* valid task_ids`?

Comment (eladkal) [reply airflow/operators/python.py:192]: Ah yeah. Ill raise a PR to fix it.

Comment (eladkal) [reply airflow/operators/python.py:192]: https://github.com/apache/airflow/pull/18623

Comment (edwardwang888) [reply airflow/operators/python.py:192]: LGTM
